### PR TITLE
cli: Add the --set flag to cochroach sql

### PR DESCRIPTION
**Pull request by chrisseto**
(multiple changes, see commit messages)

Resolves: #24496

### CODE REVIEW

**Comment by knz** [comment]
Thank you for your contribution! I am so excited we'll become able to do this.

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/cli/sql.go, line 1176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26287#-LDr2iLs8k97ujInC-NG:-LDr2iLt74FLbEzi_Moi:b8yrcq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f92a64ca7044430ca68d933ee803d1dc1753655b/pkg/cli/sql.go#L1176)):*
> ```Go
> 		// Execute any \set commands to allow setting client variables
> 		// before statement execution non-interactive mode.
> 		for i := 0; i < len(sqlCtx.setStmts); i++ {
> ```

This is not the correct place to perform this initialization.

Because `runStatements` does not use `cliState` it has nothing to gain from performing this initialization early.

Instead, move this loop into the `case cliStart`  at the start of `runInteractive()`. You can keep the passing of `conn` to `runInteractive`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26287#-:-LDr2d2V4lv_J6xMZXP9:bhfni9i)*

**Comment by chrisseto** [comment]
@knz, `handleSet` also interacts with `sqlCtx` and `cliCtx`. The referenced issue suggested that the main use case of this feature would be in conjunction with the `--execute` option, to modifying the output of `runQueryAndFormatResults`
Would it be better to move that loop and the `runStatements` loop into `cliStart`?

**Comment by knz** [comment]
>  Would it be better to move that loop and the runStatements loop into cliStart?

I wouldn't phrase it like that, this would become a discussion directed by the technical solution.

Instead consider this: users have been asking for the opportunity to run arbitrary SQL commands with `-e` even when an interactive shell is desired, as prefix to whatever the user will type in. i.e.

`cockroach sql -i -e 'select 10'` would run `select 10` and then open the interactive prompt.

(This needs the addition of a `-i` flag, but that's a detail in this conversation)

Long story short you could take the opportunity to fold the two separate functionalities (interactive and non-interactive execution) into just one function in the code, which would both simplify the code and make the tool more useful.

And then I think that would answer your concern.

Does this make sense?

**Comment by chrisseto** [comment]
Yep, I'll try and update it tonight!

**Comment by chrisseto** [comment]
@knz Updated, thanks for looking!

**Comment by knz** [comment]
Thank you! Your change LGTM.

Because the `\set` command previously did not allow a user to *unset* an option, I added an additional commit before yours to recognize the syntax `a=b`. Also I extended your change to also work with `cockroach demo`.

Thank you again for your contribution!

bors r+
